### PR TITLE
Qualcomm AI Engine Direct - Support kv_cached stories 110M llama2

### PR DESCRIPTION
**Pull request by chunit-quic**
- Add custom memory descirptor
- Add e2e example script verified with story110M in 8a8w, 16a4w
- Add qnn_llama_runner to run static LLAMA.
- Add readme
- Add slice op test
- Change RemoveClone to RemoveRedundancy
- Change SimpleADB parameter artifact to build_path and related codes
- Change multihead attentions to multiple single head.
- Move sort inputs from execute to init
- Remove split op
- Support u16 and u8 mixed-precision quantization.

### CODE REVIEW

**Comment by chunit-quic** [comment]
@cccclai 

This PR contains our own static llama and corresponding runner. We use stories 110M for verification. Based on this PR. There are two more PR will be submitted recently.
1. [unified model](https://github.com/pytorch/executorch/pull/4030)
2. llama 7b from ai-hub
About the second pr, we leaverage the pre-compiled context binary from ai-hub and export it to pte. More details will be included in the in-coming pr.

By the way. We notice some of our test cases might encounter error if we chage capture_pre_autograd_graph to torch.export. Take a test case in test_qnn_delegate.py for example (TestQNNQuantizedModel.test_qnn_backend_pixel_unshuffle_math_equivalent), it will fail at prepare_pt2e. May you kindly give us some recomendation?

**Comment by cccclai** [comment]
This is a great milestone! Thank you for putting it up. Regarding this

> By the way. We notice some of our test cases might encounter error if we change capture_pre_autograd_graph to torch.export. Take a test case in test_qnn_delegate.py for example (TestQNNQuantizedModel.test_qnn_backend_pixel_unshuffle_math_equivalent), it will fail at prepare_pt2e. May you kindly give us some recommendation?

I don't see the change from `capture_pre_autograd_graph` to `torch.export` in this pr. I assume the test still pass in this pr?

**Comment by cccclai** [comment]
Another thing is - is there a way to run the test on the host machine instead of device, like simulator?

File: backends/cadence/CMakeLists.txt [hunk 1]
```diff
@@ -27,4 +27,3 @@ set(_common_include_directories ${EXECUTORCH_ROOT}/..)
 
 add_subdirectory(${CMAKE_CURRENT_SOURCE_DIR}/hifi/operators)
 add_subdirectory(${CMAKE_CURRENT_SOURCE_DIR}/hifi/kernels)
```
Comment (cccclai) [inline backends/cadence/CMakeLists.txt:29]: I think this is fixed. Mind rebasing again?

File: backends/qualcomm/builders/op_cast.py [hunk 2]
```diff
@@ -1,57 +0,0 @@
...
-import executorch.backends.qualcomm.python.PyQnnWrapperAdaptor as PyQnnWrapper
-
-import torch
-
-from .node_visitor import NodeVisitor, register_node_visitor
-from .qnn_constants import OpCast, QNN_OP_PACKAGE_NAME_QTI_AISW
-
-
-@register_node_visitor
-class Cast(NodeVisitor):
```
Comment (cccclai) [inline backends/qualcomm/builders/op_cast.py:17]: What was the reason to remove it?

File: backends/qualcomm/passes/build_quant_io.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        else:
+            return None
+
+    def _build(self, graph_module: torch.fx.GraphModule) -> torch.fx.GraphModule:
+        # forcely update delegate node's meta['spec'] to get correct output
+        # tensor size in runtime
+        call_delegate = [
+            node
+            for node in graph_module.graph.nodes
+            if node.op == "call_function" and node.name == "executorch_call_delegate"
```
Comment (cccclai) [inline backends/qualcomm/passes/build_quant_io.py:36]: Did you expect to run this pass after to_backend? Was it because of some issues you run into?

**Comment by chunit-quic** [comment]
> This is a great milestone! Thank you for putting it up. Regarding this

Thank you for promp reply! Really appricate it. :D

> I don't see the change from `capture_pre_autograd_graph` to `torch.export` in this pr. I assume the test still pass in this pr?

Yes, we are still using `capture_pre_autograd_graph` and this pr works fine. The following is the error log if we change it to torch.export.

```bash
diff --git a/backends/qualcomm/tests/utils.py b/backends/qualcomm/tests/utils.py
index 295033e5..0e615dee 100644
--- a/backends/qualcomm/tests/utils.py
+++ b/backends/qualcomm/tests/utils.py
@@ -230,7 +230,7 @@ class TestQNN(unittest.TestCase):
         custom_quant_annotations: Tuple[Callable] = (),
         quant_dtype: QuantDtype = QuantDtype.use_8a8w,
     ) -> torch.fx.GraphModule:
-        m = torch._export.capture_pre_autograd_graph(module, inputs)
+        m = torch.export.export(module, inputs).module()

         quantizer = QnnQuantizer()
         quantizer.add_custom_quant_annotations(custom_quant_annotations)

======================================================================
ERROR: test_qnn_backend_pixel_unshuffle_math_equivalent (__main__.TestQNNQuantizedModel)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "${executorch}/backends/qualcomm/tests/test_qnn_delegate.py", line 1107, in test_qnn_backend_pixel_unshuffle_math_equivalent
    module = self.get_qdq_module(module, sample_input)
  File "${executorch}/backends/qualcomm/tests/utils.py", line 253, in get_qdq_module
    prepared(*inputs)
  File "${python3.10}/site-packages/torch/fx/graph_module.py", line 738, in call_wrapped
    return self._wrapped_call(self, *args, **kwargs)
  File "${python3.10}/site-packages/torch/fx/graph_module.py", line 316, in __call__
    raise e
  File "${python3.10}/site-packages/torch/fx/graph_module.py", line 303, in __call__
    return super(self.cls, obj).__call__(*args, **kwargs)  # type: ignore[misc]
  File "${python3.10}/site-packages/torch/nn/modules/module.py", line 1657, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "${python3.10}/site-packages/torch/nn/modules/module.py", line 1675, in _call_impl
    return forward_call(*args, **kwargs)
  File "<eval_with_key>.35", line 11, in forward
    _unsafe_view = torch.ops.aten._unsafe_view.default(activation_post_process_2, [2, 8, 3, 3]);  activation_post_process_2 = None
  File "/local2/mnt/workspace/home_speedup_link/local_lib/python3.10/site-packages/torch/_ops.py", line 670, in __call__
    return self_._op(*args, **kwargs)
RuntimeError: view size is not compatible with input tensor's size and stride (at least one dimension spans across two contiguous subspaces). Use .reshape(...) instead.

----------------------------------------------------------------------
Ran 1 test in 0.111s
```

**Comment by chunit-quic** [comment]
> Another thing is - is there a way to run the test on the host machine instead of device, like simulator?

Should be possible. We could link HTP lib to `x86_64-linux-clang/libQnnHtp.so`, and try to run on x86. Yet currently we don't have enough resource. We will add it to our TODO list.

File: backends/cadence/CMakeLists.txt (hunk 1 above)
Comment (chunit-quic) [reply backends/cadence/CMakeLists.txt:29]: Sure. Will do

File: backends/qualcomm/builders/op_cast.py (hunk 2 above)
Comment (chunit-quic) [reply backends/qualcomm/builders/op_cast.py:17]: Pardon for missleading. We change this file to [op_to.py](https://github.com/pytorch/executorch/pull/4142/files#diff-1b345d4dcc7c7035fa648723d76b7282e1c54ef06216de42084af1910a912e2c).

The reason is that, we should convert torch to_copy op to two different backend ops, cast or convert conditionally. Therefore we make a new file for it.

File: backends/qualcomm/passes/build_quant_io.py (hunk 3 above)
Comment (chunit-quic) [reply backends/qualcomm/passes/build_quant_io.py:36]: Thank you for pointing it out.
Yes, we intentionally invoke this pass after to_backend to change the io data type to non-fp type.

**Comment by chiwwang** [comment]
> > Another thing is - is there a way to run the test on the host machine instead of device, like simulator?
> 
> Should be possible. We could link HTP lib to `x86_64-linux-clang/libQnnHtp.so`, and try to run on x86. Yet currently we don't have enough resource. We will add it to our TODO list.

Actually it has been in our TODO list for a while..... 😢

File: backends/qualcomm/passes/build_quant_io.py (hunk 3 above)
Comment (cccclai) [reply backends/qualcomm/passes/build_quant_io.py:36]: I'm trying to understand the idea. It seems like foldqdq will remove all the q/dq nodes, and insert io qdq passes will insert q/dq nodes for the subgraph inside qnn backend again. In this case, why would the io data type become non-fp type?

**Comment by cccclai** [comment]
Another note, I noticed there are many passes running before to_backend for recomposing ops and the `capture_program` function control the decomp table a bit. 

[There is an experimental api `def _to_edge_transform_and_lower`](https://github.com/pytorch/executorch/blob/f32d707fd89ce64ffbaff6d9c216ac7d15c829e6/exir/program/_program.py#L862-L872) landed recently to stop decomp somes op, in case you find it helpful

File: backends/qualcomm/passes/build_quant_io.py (hunk 3 above)
Comment (chunit-quic) [reply backends/qualcomm/passes/build_quant_io.py:36]: Right, the flow you mentioned above is exactly what we have done before. Yet in LLAMA we specifically [deal with its KV IOs](https://github.com/pytorch/executorch/pull/4142/files#diff-716ed91cb062d062c9434ce96b213e0941cc24e58314de3686acfce5f53c4bb0R251). Quantizing and dequantizing are meaningless for these IOs. Therefore we keep them as quantized type all the time. 

Here is what we have done for these kind of tensors.
1.Tag IO nodes [as a quantized type](https://github.com/pytorch/executorch/pull/4142/files#diff-716ed91cb062d062c9434ce96b213e0941cc24e58314de3686acfce5f53c4bb0R340) in examples/qualcomm/llama2/llama.py:340
2. Skip insert quantized/dequantized nodes in [insert_io_qdq](https://github.com/pytorch/executorch/pull/4142/files#diff-482726e32a75c5a6f3c6439ea06a97383d6865237e8ed8a5d6193cb43b268f2dR115)
3. Set correct quantized type to [their spec](https://github.com/pytorch/executorch/pull/4142/files#diff-716ed91cb062d062c9434ce96b213e0941cc24e58314de3686acfce5f53c4bb0R316) in examples/qualcomm/llama2/llama.py:316

Feel free to let me know if anything is unclear to you. Thank you. :D

**Comment by chunit-quic** [comment]
> Another note, I noticed there are many passes running before to_backend for recomposing ops and the `capture_program` function control the decomp table a bit.
> 
> [There is an experimental api `def _to_edge_transform_and_lower`](https://github.com/pytorch/executorch/blob/f32d707fd89ce64ffbaff6d9c216ac7d15c829e6/exir/program/_program.py#L862-L872) landed recently to stop decomp somes op, in case you find it helpful

Thank you for pointing out! We will find some time to try this function in our lowering flow.

**Comment by cccclai** [comment]
Sure and no pressure at all. It's an experimental api and there is no plan to replace it with the existing api. Just would like to share in case it's helpful (as we're also looking for feedback to help the devx for backend authors)

**Review by cccclai — APPROVED** [review]

**Comment by cccclai** [comment]
> So maybe I also need [stack_trace ](https://github.com/pytorch/executorch/blob/28a45cdbe1bf1b41673b6418a09b040e1c0733e9/exir/print_program.py#L417)to identify which node we want. Is it stable?

We can give it a shot. If it's problematic, we can work Compiler team to figure out a better solution

> Do you mean we need to handle [the life cycle of the processed](https://github.com/pytorch/executorch/blob/28a45cdbe1bf1b41673b6418a09b040e1c0733e9/backends/qualcomm/runtime/QnnExecuTorchBackend.cpp#L25C21-L25C30) in the custom op?
Originally, we [load qnn context binary in init function of QnnBackend](https://github.com/pytorch/executorch/blob/28a45cdbe1bf1b41673b6418a09b040e1c0733e9/backends/qualcomm/runtime/backends/QnnContextCommon.cpp#L45) and [unload it in destroy function of QnnBackend](https://github.com/pytorch/executorch/blob/28a45cdbe1bf1b41673b6418a09b040e1c0733e9/backends/qualcomm/runtime/backends/QnnContextCommon.cpp#L18). For the life cycle of qnn context binary is decided by processed which is kept by executorch runtime framework. Is this understanding correct?

From what I understand,  every shard is a qnn context binary and it almost take all the memory in dsp. Among 4 shards, we'd need to load and unload for every inference. In the meanwhile, we'll use the spill buffer to optimize memory among shards.

If we use the custom op to shard the graph, I image we'd need to destroy the previous context binary and load the next one?

**Pull request closed without merging**